sound/soc/codecs/rt711-sdca-sdw.c

```diff
@@ -0,0 +1,394 @@
+	addr = prop->sink_ports;
+	for_each_set_bit(bit, &addr, 32) {
+		dpn[i].num = bit;
+		dpn[i].type = SDW_DPN_FULL;
+		dpn[i].simple_ch_prep_sm = true;
+		dpn[i].ch_prep_timeout = 10;
+		i++;
+	}
+
+	/* set the timeout values */
+	prop->clk_stop_timeout = 20;
+
+	/* wake-up event */
+	prop->wake_capable = 1;
+
+	return 0;
+}
+
+static int rt711_sdca_interrupt_callback(struct sdw_slave *slave,
+					struct sdw_slave_intr_status *status)
+{
+	struct rt711_sdca_priv *rt711 = dev_get_drvdata(&slave->dev);
+	int ret, stat, count = 0, retry = 3;
+	unsigned int sdca_cascade, scp_sdca_stat1, scp_sdca_stat2;
+	unsigned int buf;
+
+	dev_dbg(&slave->dev,
+		"%s control_port_stat=%x, sdca_cascade=%x", __func__,
+		status->control_port, status->sdca_cascade);
+
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &rt711->scp_sdca_stat1);
+	if (ret < 0)
+		goto io_error;
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &rt711->scp_sdca_stat2);
+	if (ret < 0)
+		goto io_error;
+
+	do {
+		/* clear flag */
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &buf);
+		if (ret < 0)
+			goto io_error;
+		if (buf & SDW_SCP_SDCA_INTMASK_SDCA_0) {
+			ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT1, SDW_SCP_SDCA_INTMASK_SDCA_0);
+			if (ret < 0)
+				goto io_error;
+		}
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &buf);
+		if (ret < 0)
+			goto io_error;
+		if (buf & SDW_SCP_SDCA_INTMASK_SDCA_8) {
+			ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT2, SDW_SCP_SDCA_INTMASK_SDCA_8);
+			if (ret < 0)
+				goto io_error;
+		}
+
+		/* check if flag clear or not */
+		ret = regmap_read(rt711->regmap, SDW_DP0_INT, &buf);
+		if (ret < 0)
+			goto io_error;
+		sdca_cascade = buf & SDW_DP0_SDCA_CASCADE;
+
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &buf);
+		if (ret < 0)
+			goto io_error;
+		scp_sdca_stat1 = buf & SDW_SCP_SDCA_INTMASK_SDCA_0;
+
+		ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT2, &buf);
+		if (ret < 0)
+			goto io_error;
+		scp_sdca_stat2 = buf & SDW_SCP_SDCA_INTMASK_SDCA_8;
+
+		stat = scp_sdca_stat1 || scp_sdca_stat2 || sdca_cascade;
+
+		count++;
+	} while (stat != 0 && count < retry);
+
+	if (stat)
+		dev_warn(&slave->dev,
+			"%s scp_sdca_stat1=0x%x, scp_sdca_stat2=0x%x\n", __func__,
+			rt711->scp_sdca_stat1, rt711->scp_sdca_stat2);
+
+	if (status->sdca_cascade)
+		mod_delayed_work(system_power_efficient_wq,
+			&rt711->jack_detect_work, msecs_to_jiffies(250));
+
+	return 0;
+
+io_error:
+	pr_err_ratelimited("IO error in %s, ret %d\n", __func__, ret);
+	return ret;
+}
+
+static struct sdw_slave_ops rt711_sdca_slave_ops = {
+	.read_prop = rt711_sdca_read_prop,
+	.interrupt_callback = rt711_sdca_interrupt_callback,
+	.update_status = rt711_sdca_update_status,
+};
+
+static int rt711_sdca_sdw_probe(struct sdw_slave *slave,
+				const struct sdw_device_id *id)
+{
+	struct regmap *regmap, *mbq_regmap;
+
+	/* Regmap Initialization */
+	mbq_regmap = regmap_init_sdw_mbq(slave, &rt711_sdca_mbq_regmap);
```

Comment:
should this be devm_ ?

Reply:
I can't find out the devm_regmap_init_sdw_mbq macro definition in regmap.h. 
It should be created in regmap.h, right?

Reply:
I see it on the sdw/sdca-integration branch:

````c
struct regmap *__devm_regmap_init_sdw(struct sdw_slave *sdw,
				 const struct regmap_config *config,
				 struct lock_class_key *lock_key,
				 const char *lock_name);
struct regmap *__devm_regmap_init_sdw_mbq(struct sdw_slave *sdw,
					  const struct regmap_config *config,
					  struct lock_class_key *lock_key,
					  const char *lock_name);
struct regmap *__devm_regmap_init_slimbus(struct slim_device *slimbus,
				 const struct regmap_config *config,
				 struct lock_class_key *lock_key,
				 const char *lock_name);
````

Reply:
I will use __devm_regmap_init_sdw_mbq instead. Thanks.

Reply:
sorry I only see this comment now @shumingfan. yes, that's a big miss on my part. Do you mind adding this patch as a fixup!  in the PR and we'll squash it with the rest?